cpp/dolfinx/fem/CoordinateElement.cpp

```diff
@@ -86,8 +95,110 @@ void CoordinateElement::compute_reference_geometry(
   assert(K.dimension(1) == this->topological_dimension());
   assert(K.dimension(2) == this->geometric_dimension());
 
-  assert(_compute_reference_geometry);
-  _compute_reference_geometry(X.data(), J.data(), detJ.data(), K.data(),
-                              num_points, x.data(), cell_geometry.data());
+  const int d = cell_geometry.rows();
+  Eigen::Tensor<double, 3, Eigen::RowMajor> phi_tensor(1, 1, d);
+  Eigen::Map<
+      Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+      phi(phi_tensor.data(), d, 1);
+
+  Eigen::Tensor<double, 4, Eigen::RowMajor> dphi_tensor(1, 1, d, _tdim);
+  Eigen::Map<
+      Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+      dphi(dphi_tensor.data(), d, _tdim);
+
+  if (_is_affine)
+  {
+    Eigen::RowVectorXd X0(_tdim);
+    X0.setZero();
+    Eigen::VectorXd x0(x.cols());
+    // Compute physical coordinates at X=0.
+    _finite_element->evaluate_reference_basis(phi_tensor, X0);
+    x0 = cell_geometry.matrix().transpose() * phi;
+
+    // Compute Jacobian and inverse
+    _finite_element->evaluate_reference_basis_derivatives(dphi_tensor, 1, X0);
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> J0(
+        _gdim, _tdim);
+    J0 = cell_geometry.matrix().transpose() * dphi;
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> K0(
+        _tdim, _gdim);
+    if (_gdim == _tdim)
+    {
+      K0 = J0.inverse();
+      detJ.fill(J0.determinant());
+    }
+    else
+    {
+      // Penrose-Moore pseudo-inverse
+      K0 = (J0.transpose() * J0).inverse() * J0.transpose();
+      detJ.fill(std::sqrt((J0.transpose() * J0).determinant()));
+    }
+
+    // Fill result for J, K and detJ
+    Eigen::Map<
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+        Jview(J.data(), _gdim * num_points, _tdim);
+    Jview = J0.replicate(num_points, 1);
+    Eigen::Map<
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>
+        Kview(K.data(), _tdim * num_points, _gdim);
+    Kview = K0.replicate(num_points, 1);
+
+    // Calculate X for each point
+    for (int ip = 0; ip < num_points; ++ip)
+      X.row(ip) = K0 * (x.row(ip).matrix().transpose() - x0);
+  }
+  else
+  {
+    // Newton's method for non-affine geometry
+    Eigen::VectorXd xk(x.cols());
+
+    for (int ip = 0; ip < num_points; ++ip)
+    {
+      Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                               Eigen::RowMajor>>
+          Jview(J.data() + ip * _gdim * _tdim, _gdim, _tdim);
+      Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic,
+                               Eigen::RowMajor>>
+          Kview(K.data() + ip * _gdim * _tdim, _tdim, _gdim);
+      Eigen::RowVectorXd Xk(_tdim);
```

Comment:
Avoid the dynamic memory allocation. Could use max size template argument,

Reply:
Yes, I thought about this. I think it might be better to template on tdim and gdim to get the best performance out of Eigen, since it will then use fixed size (stack-based) arrays. TODO.

Reply:
Anyway, I can move this one out of the loop at least